hack/compare_states.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/golang/protobuf/proto"
+	"github.com/google/go-cmp/cmp"
+	"github.com/sirupsen/logrus"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type options struct {
+	first, second gcs.Path
+	creds         string
+}
+
+// validate ensures reasonable options
+func (o *options) validate() error {
+	if o.first.String() == "" || o.second.String() == "" {
+		logrus.Fatalf("--first and --second are required and must specify directories.")
+	}
+	if !strings.HasSuffix(o.first.String(), "/") {
+		o.first.Set(o.first.String() + "/")
+	}
+	if !strings.HasSuffix(o.second.String(), "/") {
+		o.second.Set(o.second.String() + "/")
+	}
+
+	return nil
+}
+
+// gatherOptions reads options from flags
+func gatherFlagOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.Var(&o.first, "first", "First directory of state files to compare.")
+	fs.Var(&o.second, "second", "Second directory of state files to compare.")
+	fs.StringVar(&o.creds, "gcp-service-account", "", "/path/to/gcp/creds (use local creds if empty)")
+	fs.Parse(args)
+	return o
+}
+
+// gatherOptions reads options from flags
+func gatherOptions() options {
+	return gatherFlagOptions(flag.CommandLine, os.Args[1:]...)
+}
+
+func rowsAndColumns(ctx context.Context, path gcs.Path, client gcs.Client) (map[string]bool, map[string]bool, error) {
+	grid, err := updater.DownloadGrid(ctx, client, path)
+	if err != nil {
+		return nil, nil, fmt.Errorf("updater.DownloadGrid(%q): %v", path.String(), err)
+	}
+
+	rows := make(map[string]bool)
+	for _, row := range grid.GetRows() {
+		rows[row.GetName()] = true
+	}
+
+	columns := make(map[string]bool)
+	for _, column := range grid.GetColumns() {
+		columns[proto.MarshalTextString(column)] = true
+	}
+
+	return rows, columns, nil
+}
+
+func compare(ctx context.Context, first, second gcs.Path, client gcs.Client) (bool, error) {
+	var diffed bool
+	logrus.Infof("*****************************")
+	logrus.Infof("Comparing %q and %q...", first.String(), second.String())
+	firstRows, firstColumns, err := rowsAndColumns(ctx, first, client)
+	if err != nil {
+		return false, err
+	}
+	secondRows, secondColumns, err := rowsAndColumns(ctx, second, client)
+	if err != nil {
+		return false, err
+	}
+
+	logrus.Infof("Checking row names...")
+	if diff := cmp.Diff(firstRows, secondRows); diff != "" {
+		logrus.Infof("\tRows differ (-first, +second): %s", diff)
+		diffed = true
+	} else {
+		logrus.Infof("\tAll %d rows match!", len(firstRows))
+	}
+
+	logrus.Infof("Checking column data...")
+	if diff := cmp.Diff(firstColumns, secondColumns); diff != "" {
+		logrus.Infof("Columns differ (-first, +second): %s", diff)
+		diffed = true
+	} else {
+		logrus.Infof("\tAll %d columns match!", len(firstColumns))
+	}
+	return diffed, nil
+}
+
+func filenames(ctx context.Context, dir gcs.Path, client gcs.Client) []string {
+	stats := client.Objects(ctx, dir, "/", "")
+	var filenames []string
+	for {
+		stat, err := stats.Next()
+		if err != nil {
```

Comment:
Should we distinguish `iterator.Done` from other errors? https://pkg.go.dev/google.golang.org/api/iterator#pkg-variables